controller/object_store_controller.go

```diff
@@ -0,0 +1,870 @@
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+var (
+	ErrObjectStorePVCNotReady        = errors.New("PVC not ready")
+	ErrObjectStoreVolumeNotReady     = errors.New("Volume not ready")
+	ErrObjectStorePVNotReady         = errors.New("PV not ready")
+	ErrObjectStoreDeploymentNotReady = errors.New("Deployment not ready")
+	ErrObjectStoreServiceNotReady    = errors.New("Service not ready")
+	ErrObjectStoreIngressNotReady    = errors.New("Ingress not ready")
+)
+
+type ObjectStoreController struct {
+	*baseController
+
+	namespace string
+	ds        *datastore.DataStore
+	s3gwImage string
+	uiImage   string
+
+	cacheSyncs []cache.InformerSynced
+}
+
+func NewObjectStoreController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace string,
+	objectStoreImage string,
+	objectStoreUIImage string,
+) *ObjectStoreController {
+	osc := &ObjectStoreController{
+		baseController: newBaseController("object-store", logger),
+		namespace:      namespace,
+		ds:             ds,
+		s3gwImage:      objectStoreImage,
+		uiImage:        objectStoreUIImage,
+	}
+
+	ds.ObjectStoreInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    osc.enqueueObjectStore,
+			UpdateFunc: func(old, cur interface{}) { osc.enqueueObjectStore(cur) },
+			DeleteFunc: osc.enqueueObjectStore,
+		},
+	)
+	osc.cacheSyncs = append(osc.cacheSyncs, ds.ObjectStoreInformer.HasSynced)
+
+	return osc
+}
+
+func (osc *ObjectStoreController) Run(workers int, stopCh <-chan struct{}) {
+	osc.logger.Info("Starting Longhorn Object Store Controller")
+	defer osc.logger.Info("Shut down Longhorn Object Store Controller")
+	defer osc.queue.ShutDown()
+
+	if !cache.WaitForNamedCacheSync("longhorn object stores", stopCh, osc.cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(osc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (osc *ObjectStoreController) worker() {
+	for osc.processNextWorkItem() {
+	}
+}
+
+func (osc *ObjectStoreController) processNextWorkItem() bool {
+	key, quit := osc.queue.Get()
+	if quit {
+		return false
+	}
+	defer osc.queue.Done(key)
+
+	err := osc.syncObjectStore(key.(string))
+	if err == nil {
+		osc.queue.Forget(key)
+	} else if osc.queue.NumRequeues(key) < maxRetries {
+		osc.logger.WithError(err).Errorf("error syncing object store: \"%v\", retrying", err)
+		osc.queue.AddRateLimited(key)
+	} else {
+		utilruntime.HandleError(err)
+		osc.logger.WithError(err).Errorf("error syncing object store: \"%v\", giving up", err)
+		osc.queue.Forget(key)
+	}
+
+	return true
+}
+
+func (osc *ObjectStoreController) enqueueObjectStore(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Failed to get key for %v: %v", obj, err))
```

Comment:
Sugestting to lowercase the first letter of the error message :
```suggestion
		utilruntime.HandleError(fmt.Errorf("failed to get key for %v: %v", obj, err))
```